File: SummaryLasso/R/transPEN.R
`transPEN` <- function(summaryZ, summaryBetas = NULL, SSvec = NULL, Nvec, plinkLDList, NumIter = 100, breaking = 1,
  numChrs=22, ChrIndexBeta = 0, Init_summaryBetas= NULL, Zscale = 1, RupperVal = NULL,
  tuningMatrix = NULL, penalty=c("Lasso","LassowII","LassowIII","transw","mixLOG"),
  taufactor = c(1/25, 1, 10), llim_length = 10, subtuning = 50, Lambda_limit = c(0.5,0.9), 
  Lenlam_singleTrait = 200, dfMax = NULL, IniBeta = 0, inverseTuning = 0, outputAll = 0, warmStart = 1 , orderseq = 1, basew = NULL, tau = 0.0, alpha = 1, singleStart = 1, weightN = 0){
 

  #if(Zscale!=1){stop("Tuning values set-up for multiple traits analysis requires Zscale=1.")}

  if(inverseTuning==1){
    orderD = "FALSE"
  }else{
    orderD = "TRUE"
  }
  Nq = length(Nvec)
  
  
  
  if(is.null(summaryBetas)){
    summaryBetas = matrix(0, nrow(summaryZ), Nq)
    SDvec = matrix(0, nrow(summaryZ), Nq)
  
    for(ii in 1:Nq){
      summaryBetas[,ii] = summaryZ[,ii]/sqrt(Nvec[ii])
      SDvec[,ii] = 1/sqrt(Nvec[ii])
    }
    rownames(summaryBetas) = rownames(summaryZ)
    SSvec = matrix(0,1,1)
  }else{
    SDvec = matrix(0,1,1)
  }
  
  if(!is.null(Init_summaryBetas)){
    IniBeta = 1
  }else{
    Init_summaryBetas = matrix(0, nrow(summaryBetas), Nq)
  }
   

  
  sumdiffbetas = apply(summaryBetas,1,abs_diff)

  

  if(is.null(dfMax)){
    dfMax = ceiling(0.7*nrow(summaryBetas))
  }
  
  autoTuning = 0
  if(is.null(tuningMatrix)){
    autoTuning = 1
    medianval = median(apply(abs(summaryBetas),1,sum),na.rm=T)
    tauvec = sort(medianval*taufactor)
    
    Lambda_limit = quantile(abs(summaryZ[,1]), Lambda_limit)

    tuningMatrix = Tuning_setup_group_only(tauvec, subtuning, Lambda_limit, Lenlam_singleTrait, llim_length, medianval)
  }            
     

  inv_summaryBetas = 0

  
  Betaindex = c(1:nrow(summaryBetas))-1
  SNPnames = rownames(summaryBetas)


  q_nrowIndexMatrix = q_ldJ = q_IndJ = q_ldvec = c()
  
  for(iiIndex in 1:Nq){
    ldJ = PlinkLD_transform(plinkLDList[[iiIndex]], SNPnames)

    JidMatrix = matrix(,nrow(ldJ),2)
    mat1 = match(ldJ[,1],SNPnames)
    mat2 = match(ldJ[,2],SNPnames)

    JidMatrix[,1] = Betaindex[mat1]
    JidMatrix[,2] = Betaindex[mat2]

    ldJ[,1] = JidMatrix[,1]
    ldJ[,2] = JidMatrix[,2]

    od = order(JidMatrix[,1],JidMatrix[,2], decreasing=F)
    ldJ = ldJ[od,]
  
    wind = which(! Betaindex %in% ldJ[,1])

    IndJ = -1

    if(length(wind) > 0){
      IndJ = Betaindex[wind]
    }

    Counts = table(ldJ[,1])
    NumSNP = length(Counts)
  
    IndexS = c(0,cumsum(Counts)[-NumSNP])
    IndexE = cumsum(Counts)-1

    IndexMatrix = matrix(,NumSNP,3)
    IndexMatrix[,1] = as.numeric(names(Counts))

    nrow_IndexMatrix = NumSNP
    ncol_IndexMatrix = ncol(IndexMatrix)

    IndexMatrix[,2] = IndexS
    IndexMatrix[,3] = IndexE

   ld_cols <- ncol(ldJ)


    ldvec <- ldJ[, 3]


    ldJ = ldJ[,2]

    if(iiIndex==1){
      All_IndexMatrix = IndexMatrix
      All_ldJ = ldJ
      All_IndJ = IndJ
      All_ldvec = ldvec
      All_length_wind = length(wind)
    }else{
      All_IndexMatrix = rbind(All_IndexMatrix, IndexMatrix)
      All_ldJ = c(All_ldJ,ldJ)
      All_IndJ = c(All_IndJ,IndJ)
      All_ldvec = c(All_ldvec,ldvec)
      All_length_wind = c(All_length_wind,length(wind))
    }
    

    q_nrowIndexMatrix = c(q_nrowIndexMatrix,NumSNP)
    q_ldJ = c(q_ldJ, length(ldJ))
    q_IndJ = c(q_IndJ, length(IndJ))
    q_ldvec = c(q_ldvec, length(ldvec))
    rm(ldJ,IndJ,ldvec,wind)
  }

  if(is.null(RupperVal)){
    RupperVal = ceiling(max(abs(summaryBetas),na.rm=T)*50)
  }

  P = nrow(summaryBetas)
  Q = ncol(summaryBetas)
  
  if(nrow(tuningMatrix)>1){
    NumTuning = nrow(tuningMatrix)
  }
  
  if(warmStart==1){
    if(Nq>1){
      if(orderseq==1){
        if(penalty=="Lasso"){
          id = order(tuningMatrix[,1],1/tuningMatrix[,2],decreasing=orderD)
        }else{
          id = order(tuningMatrix[,1],tuningMatrix[,2],decreasing=orderD)
        }
        useI = 1
      }
      if(orderseq==2){
        if(penalty=="Lasso"){
          id = order(1/tuningMatrix[,2],tuningMatrix[,1],decreasing=orderD)
        }else{
          id = order(tuningMatrix[,2],tuningMatrix[,1],decreasing=orderD)
        }
        useI = 2
      }
      tuningMatrix = tuningMatrix[id,]
      groupvec = apply(tuningMatrix[,c(useI),drop=F],1,paste0,collapse="")
      Tgroupvec = table(groupvec)
      Ugroup = unique(groupvec)
      mat = match(Ugroup, names(Tgroupvec))
      Tgroupvec = Tgroupvec[mat]
      if(length(Ugroup)==1){
        StartVec = c(1)
      }else{
        StartVec = c(1, (cumsum(Tgroupvec)[1:(length(cumsum(Tgroupvec))-1)]+1))
      }
    }else{
      if(orderseq==1){
        id = order(tuningMatrix[,1],decreasing=orderD)
      }
      if(orderseq==2){
        id = order(tuningMatrix[,3],decreasing=orderD)
      }
      tuningMatrix = tuningMatrix[id,]
      StartVec = 1
    }
    tuningMatrix = cbind(tuningMatrix,0)
    tuningMatrix[StartVec,ncol(tuningMatrix)] = 1
  }else{
    tuningMatrix = cbind(tuningMatrix,0)
  }
  if(singleStart==1){
    tuningMatrix[,ncol(tuningMatrix)] = 0
    tuningMatrix[1,ncol(tuningMatrix)] = 1
  }

  nrow_All_IndexMatrix = nrow(All_IndexMatrix)
  ncol_All_IndexMatrix = ncol(All_IndexMatrix)

  ncolBetaMatrix = P*Q
  dims = rep(0,20)
  dims[1] = NumTuning
  dims[2] = P
  dims[3] = NumIter
  dims[4] = breaking
  dims[5] = nrow_All_IndexMatrix
  dims[6] = ncol_All_IndexMatrix
  dims[7] = 0  #NumInd
  dims[8] = Zscale
  dims[9] = nrow(tuningMatrix)
  dims[10] = ncol(tuningMatrix)
  dims[11] = Q
  dims[12] = ncolBetaMatrix
  dims[13] = outputAll
  dims[14] = dfMax
  dims[15] = warmStart
  dims[16] = IniBeta
  dims[17] = nrow(SDvec)
  dims[18] = ncol(SDvec)
  dims[19] = nrow(SSvec)
  dims[20] = ncol(SSvec)
  
  if(is.null(basew)){
    basew = max(sumdiffbetas,na.rm=T)
  }
  doublevec = c(RupperVal, basew, tau, alpha, weightN)

  Numitervec = rep(0, NumTuning)
  Dfq1 = rep(0, NumTuning)
  BetaMatrix = matrix(0, NumTuning, ncolBetaMatrix)
  
  
  

  Z = .C("transPEN", as.double(t(summaryBetas)),as.integer(All_ldJ), as.integer(dims), Numitervec=as.integer(Numitervec),
   as.integer(t(All_IndexMatrix)), as.integer(All_IndJ),
   as.double(All_ldvec), as.double(inv_summaryBetas), as.integer(ChrIndexBeta), as.double(doublevec),
   as.double(t(Init_summaryBetas)), as.double(t(SDvec)),
   as.double(t(tuningMatrix)), BetaMatrix = as.double(t(BetaMatrix)),
   penalty, as.integer(q_nrowIndexMatrix), as.integer(q_ldJ),as.integer(q_IndJ),as.integer(q_ldvec), as.integer(All_length_wind), Dfq1 = as.integer(Dfq1), as.double(t(SSvec)),PACKAGE="SummaryLasso")




  BetaMatrix = matrix(Z$BetaMatrix, nrow = NumTuning, ncol = ncolBetaMatrix, byrow = TRUE)
  tuningMatrix = tuningMatrix[,-ncol(tuningMatrix)]
  
  if(autoTuning){
    tuningMatrix[c(1:Lenlam_singleTrait), c(2:3)] = NA
  } 
  colnames(BetaMatrix) = paste0(rep(SNPnames, times = Q),".trait", rep(c(1:Q), each = P))
   
  Numitervec = Z$Numitervec
  Dfq1 = Z$Dfq1
   
  if(outputAll==0){
    convergeIndex = which(Numitervec > 0)
    Numitervec = Numitervec[convergeIndex]
    BetaMatrix = BetaMatrix[convergeIndex,]
    tuningMatrix = tuningMatrix[convergeIndex,]
  }
   
  ll = list(BetaMatrix = BetaMatrix, Numitervec = Numitervec, tuningMatrix = tuningMatrix,Dfq1=Dfq1)
  return(ll)
   
}

File: SummaryLasso/src/transPEN.c
/*
 *  IterativeSNPC.c
 *
 *  Created by Ting-Huei Chen on Feb 06, 2015.
 *
 *  Last updated by Ting-Huei Chen on Feb 25, 2018.
 */
#include <R_ext/Lapack.h>
#include <R_ext/Applic.h>
#include <R_ext/PrtUtil.h>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>
#include <R.h>
#include <Rmath.h>
#include "utility.h"
#include "transPEN.h"



/*********************************************************************
 *
 * 
 *
 *********************************************************************/
 
void transPEN(double* RsummaryBetas,int* All_ldJ, int* dims, int* Numitervec,
     int* RAll_IndexMatrix, int* All_IndJ,
     double* All_ldvec, double* inv_summaryBetas, int*ChrIndexBeta, double*doublevec,
     double* RInit_summaryBetas, double* RSDvec,
     double* RtuningMatrix, double* RBetaMatrix, char **penalty_, int* q_nrowIndexMatrix, int* q_ldJ, int* q_IndJ, int* q_ldvec, int* All_length_wind, int* Dfq1, double* RSSvec)
{
    int NumTuning = dims[0], P = dims[1];
    int nrow_All_IndexMatrix = dims[4], ncol_All_IndexMatrix = dims[5];
    int **All_IndexMatrix;
    int ncol_tuningMatrix = dims[9], Q = dims[10], ncolBetaMatrix = dims[11];
    int nrow_SDvec = dims[16], ncol_SDvec = dims[17];
    int nrow_SSvec = dims[18], ncol_SSvec = dims[19];
    double **BetaMatrix, **tuningMatrix, **summaryBetas, **SDvec, **Init_summaryBetas;
    double **SSvec;
    
    /* reorganize vector into matrix */
    reorg(RsummaryBetas, &summaryBetas, P, Q);
    reorg(RInit_summaryBetas, &Init_summaryBetas, P, Q);
    reorg(RSDvec, &SDvec, nrow_SDvec, ncol_SDvec);
    reorg(RSSvec, &SSvec, nrow_SSvec, ncol_SSvec);
    reorg(RBetaMatrix, &BetaMatrix, NumTuning, ncolBetaMatrix);
    
    reorg(RtuningMatrix, &tuningMatrix, NumTuning, ncol_tuningMatrix);
    reorg_int(RAll_IndexMatrix, &All_IndexMatrix, nrow_All_IndexMatrix, ncol_All_IndexMatrix);
    //Rprintf("check=%d\n",1);
    transPENC(summaryBetas, All_ldJ, dims, Numitervec,
     All_IndexMatrix, All_IndJ,
     All_ldvec, inv_summaryBetas, ChrIndexBeta, doublevec,
     Init_summaryBetas, SDvec, 
     tuningMatrix, BetaMatrix, penalty_, q_nrowIndexMatrix, q_ldJ, q_IndJ, q_ldvec, All_length_wind, Dfq1, SSvec);
} 


void transPENC(double** summaryBetas, int* All_ldJ, int* dims, int* Numitervec,
     int** All_IndexMatrix, int* All_IndJ,
     double* All_ldvec, double* inv_summaryBetas, int*ChrIndexBeta, double*doublevec,
     double** Init_summaryBetas, double** SDvec, 
     double**tuningMatrix, double** BetaMatrix,
     char **penalty_,int* q_nrowIndexMatrix, int* q_ldJ, int* q_IndJ, int* q_ldvec, int* All_length_wind, int* Dfq1, double** SSvec)
{
  
    int NumTuning = dims[0], P = dims[1], NumIter = dims[2], breaking = dims[3];
    int niter, t1, i, j,j1, j2, found, **skipb;
    int NumSNP, NumInd = dims[6];
    int Zscale = dims[7];

    int ncol_tuningMatrix = dims[9], Q = dims[10], k, k1, k2, cindex, warmStart = dims[14], IniBeta = dims[15];
    int OutputForPath = dims[12], dfMax = dims[13], df_q1 = 0.0;
    char *penalty=penalty_[0];
    
    int start_IndJ, start_ldJ, start_ldvec, start_nrowIndexMatrix, ktemp;

    
    double tmp0, tmp1, tmp2, *tmpbetas, threshold = 0.0, bj_bar = 0.0, epsilon = 0.0001;
    double *sumDiffBetas, *sumBetas, *ratio, upperVal = doublevec[0], basew = doublevec[1];
    double lambda0 = 0.0, tau = doublevec[2], alpha = doublevec[3], lambda1 = 0.0;
    double *sumBetas_after, weightN = doublevec[4];
    

    
    /* allocate memory */
    tmpbetas = (double *)calloc(P, sizeof(double));
    sumDiffBetas = (double *)calloc(P, sizeof(double));
    sumBetas = (double *)calloc(P, sizeof(double));
    sumBetas_after = (double *)calloc(P, sizeof(double));
    ratio = (double *)calloc(P, sizeof(double));
    
    GetRNGstate();
    
    double **jointBmatrix, **tempBmatrix;
    
    jointBmatrix = (double **) malloc(P * sizeof(double*));
    tempBmatrix = (double **) malloc(P * sizeof(double*));
    skipb = (int **) malloc(P * sizeof(int*));
  
    jointBmatrix[0] = (double *) calloc(P*Q, sizeof(double));
    if(jointBmatrix[0] == NULL){ error("fail to allocate memory of jointBmatrix"); }
    for(j=0; j<P; j++){  
      jointBmatrix[j] = jointBmatrix[0] + j*Q; 
    }
    
    tempBmatrix[0] = (double *) calloc(P*Q, sizeof(double));
    if(tempBmatrix[0] == NULL){ error("fail to allocate memory of tempBmatrix"); }
    for(j=0; j<P; j++){  
      tempBmatrix[j] = tempBmatrix[0] + j*Q; 
    }
    
    skipb[0] = (int *) calloc(P*Q, sizeof(int));
    if(skipb[0] == NULL){ error("fail to allocate memory of skipb"); }
    for(j=0; j<P; j++){  
      skipb[j] = skipb[0] + j*Q; 
    }

    for(i=0; i<P; i++){  
      sumDiffBetas[i] = 0.0;
      for(j=0; j<Q; j++){  
        jointBmatrix[i][j] = 0.0;
        tempBmatrix[i][j] = 0.0;
        skipb[i][j] = 0;
      } 
    }
    
    //printf("(ncol_tuningMatrix-1)=%d\n",(ncol_tuningMatrix-1));
    //Rprintf("tuningMatrix[0][0]=%f\n",tuningMatrix[0][0]);
    //Rprintf("tuningMatrix[0][1]=%f\n",tuningMatrix[0][1]);
    //Rprintf("(pow(3.0,2.0)=%f\n",pow(3.0,2.0));

    
    for(t1=0; t1<NumTuning; t1++){
      //printf("t1=%d\n",t1);
      Numitervec[t1] = 0;
      
      lambda0 = tuningMatrix[t1][0];
      lambda1 = tuningMatrix[t1][1];
      
      
      if (strcmp(penalty,"LassowII")==0){
        alpha = tuningMatrix[t1][1];
      }
      
      if(warmStart==0){
        for(i=0; i<P; i++){
          for(j=0; j<Q; j++){
            jointBmatrix[i][j] = 0.0;
            tempBmatrix[i][j] = 0.0;
            skipb[i][j] = 0;
          }
          sumDiffBetas[i] = 0.0;
          sumBetas_after[i] = 0.0;
          sumBetas[i] = 0.0;
          ratio[i] = 1.0;
        }
      }
      //Rprintf("check=%d\n",2);
      if(t1 > 0){
        if(warmStart==1 && tuningMatrix[t1][(ncol_tuningMatrix-1)]==0 && tuningMatrix[t1-1][(ncol_tuningMatrix-1)]==0){
          if(Numitervec[t1-1] <= 0){
            continue;
          }
        }
      }
      //Rprintf("t1=%d\n",t1);
      //Rprintf("warmStart=%d\n",warmStart);
      //Rprintf("tuningMatrix[t1][(ncol_tuningMatrix-1)]=%f\n",tuningMatrix[t1][(ncol_tuningMatrix-1)]);
      if(warmStart==1 && tuningMatrix[t1][(ncol_tuningMatrix-1)]==1){
        //Rprintf("warm t1=%d\n",t1);
        for(i=0; i<P; i++){
          for(j=0; j<Q; j++){
            if(IniBeta==1){
              jointBmatrix[i][j] = Init_summaryBetas[i][j];
            }else{
              jointBmatrix[i][j] = 0.0;
            }
            tempBmatrix[i][j] = 0.0;
            skipb[i][j] = 0;
          }
          sumDiffBetas[i] = 0.0;
          sumBetas[i] = 0.0;
          sumBetas_after[i] = 0.0;
          ratio[i] = 1.0;
        }
      }
      //Rprintf("check=%d\n",3);
      cindex = 0;
      for(niter=0; niter<NumIter; niter++){
        //printf("niter=%d\n",niter);
        for(k1=0; k1<Q; k1++){
          //printf("k1=%d\n",k1);
          NumSNP = q_nrowIndexMatrix[k1];
          start_IndJ = 0;
          start_ldJ = 0;
          start_ldvec = 0;
          start_nrowIndexMatrix = 0;
          if(k1 > 0){
            for(ktemp=0; ktemp<k1; ktemp++){
              start_IndJ = start_IndJ + q_IndJ[ktemp];
              start_ldJ = start_ldJ + q_ldJ[ktemp];
              start_ldvec = start_ldvec + q_ldvec[ktemp];
              start_nrowIndexMatrix = start_nrowIndexMatrix + q_nrowIndexMatrix[ktemp];
            }
          }
          NumInd = All_length_wind[k1];
          //Rprintf("check=%d\n",4);
          if(NumInd!=0){
            for(j1=0; j1<NumInd; j1++){
              j = All_IndJ[j1+start_IndJ];
              if (strcmp(penalty,"Lasso")==0){
                threshold = lambda0 + pow(sumDiffBetas[j], alpha);
              }
              if (strcmp(penalty,"LassowII")==0){
                if(sumBetas[j] == 0){
                  threshold = lambda0 + (basew + sumDiffBetas[j])*alpha;
                }else{
                  threshold = lambda0 + sumDiffBetas[j]*alpha;
                }
              }
              if (strcmp(penalty,"LassowIII")==0){
                threshold = lambda0 + lambda1*(sumDiffBetas[j]+tau)/(sumBetas[j]+tau);
              }
              if (strcmp(penalty,"transw")==0){
                threshold = lambda0 + lambda1*(sumDiffBetas[j]+basew)/(sumBetas[j]+tau);
              }
              if (strcmp(penalty,"transwII")==0){
                threshold = lambda0 + lambda1*(ratio[j]+basew);
              }
              if (strcmp(penalty,"transwIII")==0){
                threshold = lambda0 + lambda1*(pow(ratio[j],basew));
              }
              if (strcmp(penalty,"ratiow")==0){
                threshold = lambda0 + (lambda1/(pow((1-ratio[j]),alpha) + tau));
              }
              if (strcmp(penalty,"ratiowS")==0){
                threshold = lambda0 + lambda1*(sumDiffBetas[j] + tau)/(sumBetas_after[j] + tau);
              }
              
              if (strcmp(penalty,"mixLOG")==0){
                  threshold = lambda0 + lambda1*(1/(sumBetas[j] + tau));
              }
              if (strcmp(penalty,"mixLOG_after")==0){
                  threshold = lambda0 + lambda1*(1/(sumBetas_after[j] + tau));
              }
              if (strcmp(penalty,"mixLOGsq")==0){
                  threshold = lambda0 + lambda1*(1/(sumBetas_after[j] + tau));
              }
              
              
              bj_bar = summaryBetas[j][k1];
              if(bj_bar!=0.0){
                if (strcmp(penalty,"ratiowII")==0){
                  threshold = lambda0 + (lambda1/(pow((1-ratio[j]),alpha) + tempBmatrix[j][k1] + tau));
                }
                if (strcmp(penalty,"ratiowIII")==0){
                  threshold = (lambda1/(pow((1-ratio[j]),alpha) + tempBmatrix[j][k1] + tau));
                }
                
                if(Zscale==1){
                  threshold = threshold*SDvec[j][k1];
                }else{
                  threshold = threshold/SSvec[j][k1];
                }
                if(bj_bar > threshold){
                  jointBmatrix[j][k1] = bj_bar - threshold;
                }else if(bj_bar < -threshold){
                  jointBmatrix[j][k1] = bj_bar + threshold;
                }else{
                  jointBmatrix[j][k1] = 0.0;
                }
              }
              if(summaryBetas[j][k1]*jointBmatrix[j][k1]<0){
                Rprintf("summaryBetas[j]=%d\n",j);
                Rprintf("summaryBetas[k1]=%d\n",k1);
                Rprintf("summaryBetas[j][k1]=%e\n",summaryBetas[j][k1]);
                Rprintf("jointBmatrix[j][k1]=%e\n",jointBmatrix[j][k1]);
                error("sign inverse");
              }
            }
          }
          //Rprintf("check=%d\n",5);
          for(j1=0; j1<NumSNP; j1++){
            //Rprintf("j1=%d\n",j1);
            j = All_IndexMatrix[j1+start_nrowIndexMatrix][0];
            if (strcmp(penalty,"Lasso")==0){
              threshold = lambda0 + pow(sumDiffBetas[j], alpha);
            }
            if (strcmp(penalty,"LassowII")==0){
              if(sumBetas[j] == 0){
                threshold = lambda0 + (basew + sumDiffBetas[j])*alpha;
              }else{
                threshold = lambda0 + sumDiffBetas[j]*alpha;
              }
            }
            if (strcmp(penalty,"LassowIII")==0){
              threshold = lambda0 + lambda1*(sumDiffBetas[j]+tau)/(sumBetas[j]+tau);
            }
            if (strcmp(penalty,"transw")==0){
              threshold = lambda0 + lambda1*(sumDiffBetas[j]+basew)/(sumBetas[j]+tau);
            }
            //printf("lambda0=%f\n",lambda0);
            //printf("lambda1=%f\n",lambda1);
            //printf("sumDiffBetas[j]=%f\n",sumDiffBetas[j]);
            //printf("basew=%f\n",basew);
            //printf("tau=%f\n",tau);
            //printf("sumBetas[j]=%f\n",sumBetas[j]);
            //printf("threshold=%f\n",threshold);
            if (strcmp(penalty,"transwII")==0){
              threshold = lambda0 + lambda1*(ratio[j]+basew);
            }
            if (strcmp(penalty,"transwIII")==0){
              threshold = lambda0 + lambda1*(pow(ratio[j],basew));
            }
            if (strcmp(penalty,"ratiow")==0){
              threshold = lambda0 + (lambda1/(pow((1-ratio[j]),alpha) + tau));
            }
            if (strcmp(penalty,"ratiowS")==0){
                threshold = lambda0 + lambda1*(sumDiffBetas[j] + tau)/(sumBetas_after[j] + tau);
            }
            if (strcmp(penalty,"mixLOG")==0){
              threshold = lambda0 + lambda1*(1/(sumBetas[j] + tau));
            }
            if (strcmp(penalty,"mixLOG_after")==0){
              threshold = lambda0 + lambda1*(1/(sumBetas_after[j] + tau));
            }
              
            if(skipb[j][k1]==0){
              if(summaryBetas[j][k1]!=0.0){
                tmp0 = 0.0;
                if(Zscale==1){
                  for(j2=All_IndexMatrix[j1+start_nrowIndexMatrix][1]; j2<(All_IndexMatrix[j1+start_nrowIndexMatrix][2]+1); j2++){
                  tmp0 = tmp0 + All_ldvec[j2+start_ldvec]*jointBmatrix[All_ldJ[j2+start_ldJ]][k1];
                  }
                }else{
                  for(j2=All_IndexMatrix[j1+start_nrowIndexMatrix][1]; j2<(All_IndexMatrix[j1+start_nrowIndexMatrix][2]+1); j2++){
                  tmp0 = tmp0 + sqrt(SSvec[j][k1])*sqrt(SSvec[All_ldJ[j2+start_ldJ]][k1])*All_ldvec[j2+start_ldvec]*jointBmatrix[All_ldJ[j2+start_ldJ]][k1];
                  }
                }
          
                if(Zscale==1){
                  bj_bar = (summaryBetas[j][k1] - tmp0);
                }else{
                  bj_bar = (SSvec[j][k1]*summaryBetas[j][k1] - tmp0)/SSvec[j][k1];
                }
                //printf("threshold=%f\n",threshold);
                //Rprintf("pow(sumDiffBetas[j], alpha)=%f\n",pow(sumDiffBetas[j], alpha));
                //Rprintf("check=%d\n",6);
                if(fabs(bj_bar)>upperVal){
                  if(breaking==1){
                    Numitervec[t1] = -1;
                    break;
                  }else{
                    bj_bar = 0.0;
                    skipb[j][k1] = 1;
                  }
                }
                if (strcmp(penalty,"ratiowII")==0){
                  threshold = lambda0 + (lambda1/(pow((1-ratio[j]),alpha) + tempBmatrix[j][k1] + tau));
                }
                if (strcmp(penalty,"ratiowIII")==0){
                  threshold = (lambda1/(pow((1-ratio[j]),alpha) + tempBmatrix[j][k1] + tau));
                }
                //Rprintf("check=%d\n",7);
                if(Zscale==1){
                  threshold = threshold*SDvec[j][k1];
                }else{
                  threshold = threshold/SSvec[j][k1];
                }
                if(bj_bar > threshold){
                  jointBmatrix[j][k1] = bj_bar - threshold;
                }else if(bj_bar < -threshold){
                  jointBmatrix[j][k1] = bj_bar + threshold;
                }else{
                  jointBmatrix[j][k1] = 0.0;
                }
              }else{
                jointBmatrix[j][k1] = 0.0;
              }  
            }else{
              jointBmatrix[j][k1] = 0.0;
            }
            //Rprintf("check=%d\n",8);
            //printf("jointBmatrix=%f\n",jointBmatrix[j][k1]);
          }
        }
        //Rprintf("check=%d\n",9);
       /**
        * check convergence
        */
              
        found = 0;
        df_q1 = 0;
        for(j=0; j<P; j++){
          for(k1=0; k1<Q; k1++){
            if(fabs(jointBmatrix[j][k1])>upperVal){
              skipb[j][k1] = 1;
              //printf("skipb%d\n",skipb[j][k1]);
            }
            if(k1==0){
              if(fabs(jointBmatrix[j][k1])!=0){
                df_q1 = df_q1 + 1;
              }
              if(df_q1 > dfMax){
                cindex = 1;
                //printf("df_q1=%d\n",df_q1);
                break;
              }
            }
            if(fabs(tempBmatrix[j][k1] - jointBmatrix[j][k1]) > epsilon){
              found = 1;
              break;
            }
          } 
        }
        //printf("df_q1=%d\n",df_q1);
        //printf("dfMax=%d\n",dfMax);
        if(cindex == 1){
          Numitervec[t1] = -2;
          //printf("Numitervec=%d\n",Numitervec[t1]);
          break;
        }
        //printf("found=%d\n",found);
        if(found==0){
          k1=0;
          df_q1 = 0;
          for(i=0; i<Q; i++){
            for(j=0; j<P; j++){
              BetaMatrix[t1][k1] = jointBmatrix[j][i];
              if(i==0){
                if(fabs(jointBmatrix[j][i])!=0){
                  df_q1 = df_q1 + 1;
                }
              }
              k1 = k1 + 1;
            } 
          } 
          Numitervec[t1] = (niter+1);
          Dfq1[t1] = df_q1;
          break;
        }
        for(j=0; j<P; j++){
          tmp1 = 0.0;
          tmp2 = 0.0;
          for(k=0; k<Q; k++){
            tempBmatrix[j][k] = jointBmatrix[j][k];
            if(weightN==0){
              tmp1 = tmp1 + fabs(jointBmatrix[j][k]);
            }else{
              if(k==0){tmp1 = tmp1 + weightN*fabs(jointBmatrix[j][k]);}
              if(k==1){tmp1 = tmp1 + (1-weightN)*fabs(jointBmatrix[j][k]);}
            }
            tmp2 = tmp2 + jointBmatrix[j][k];
          }
          sumBetas[j] = tmp1;
          sumBetas_after[j] = fabs(tmp2);
          
          tmp0 = 0.0;
          for(k1=0; k1<(Q-1); k1++){
            for(k2=(k1+1); k2<Q; k2++){
              tmp0 = tmp0 + fabs(jointBmatrix[j][k1]-jointBmatrix[j][k2]);
            }
          }
          sumDiffBetas[j] = tmp0;
          
          if (strcmp(penalty,"ratiow")==0){
            if(sumBetas[j]==0){
              ratio[j] = 1;
            }else{
              ratio[j] = sumDiffBetas[j]/sumBetas[j];
            }
          }
    
          if (strcmp(penalty,"ratiowS")==0){
            if(sumBetas_after[j]==0){
              ratio[j] = 1;
            }else{
              ratio[j] = sumDiffBetas[j]/sumBetas_after[j];
            }
          }
        }
        //printf("Numitervec1=%d\n",Numitervec[0]);
        //printf("(NumIter-1)=%d\n",(NumIter-1));
        //printf("OutputForPath=%d\n",OutputForPath);

        if(OutputForPath==1 && niter==(NumIter-1)){
          //printf("OutputForPathENTER=%d\n",OutputForPath);
          k1=0;
          for(i=0; i<Q; i++){
            for(j=0; j<P; j++){
              BetaMatrix[t1][k1] = jointBmatrix[j][i];
              k1 = k1 + 1;
            } 
          } 
        }
      }
    } 
    PutRNGstate();
    
    free(jointBmatrix[0]);
    free(tempBmatrix[0]);
    free(skipb[0]);
    free(jointBmatrix);
    free(tempBmatrix);
    free(skipb);
    free(tmpbetas);
    free(sumDiffBetas);

    
 }

 
